synapse/lib/base.py

```diff
@@ -0,0 +1,571 @@
+            fevt.set()
+        self._syn_funcs.clear()
+        del self._fini_funcs[:]
+        return 0
+
+    async def waitfini(self, timeout=None):
+        '''
+        Wait for the event bus to fini()
+
+        Example:
+
+            bus.waitfini(timeout=30)
+
+        '''
+
+        if self.isfini:
+            return True
+
+        if self.finievt is None:
+            self.finievt = asyncio.Event(loop=self.loop)
+
+        try:
+            await asyncio.wait_for(self.finievt.wait(), timeout, loop=self.loop)
+        except asyncio.TimeoutError:
+            return None
+        return True
+
+    def main(self):
+        '''
+        Helper function to block until shutdown ( and handle ctrl-c and SIGTERM).
+
+        Examples:
+            Run a event bus, wait until main() has returned, then do other stuff::
+
+                foo = EventBus()
+                foo.main()
+                dostuff()
+
+        Notes:
+            This does fire a 'ebus:main' event prior to entering the
+            waitfini() loop.
+
+        Returns:
+            None
+        '''
+        assert self.loop == s_glob.plex.loop
+
+        doneevent = threading.Event()
+        self.onfini(doneevent.set)
+
+        async def sighandler():
+            print('Caught SIGTERM, shutting down')
+            await self.fini()
+
+        def handler():
+            asyncio.run_coroutine_threadsafe(sighandler(), loop=self.loop)
+
+        try:
+            s_glob.plex.loop.add_signal_handler(signal.SIGTERM, handler)
+        except Exception as e:  # pragma: no cover
+            logger.exception('Unable to register SIGTERM handler.')
+
+        async def asyncmain():
+            await self.fire('ebus:main')
+
+        asyncio.run_coroutine_threadsafe(asyncmain(), loop=self.loop)
+
+        try:
+            doneevent.wait()
+
+        except KeyboardInterrupt as e:
+            print('ctrl-c caught: shutting down')
+
+        finally:
+            # Avoid https://bugs.python.org/issue34680 by removing handler before closing down
+            s_glob.plex.loop.remove_signal_handler(signal.SIGTERM)
+
+            asyncio.run_coroutine_threadsafe(self.fini(), loop=self.loop)
+
+    def waiter(self, count, *names):
+        '''
+        Construct and return a new Waiter for events on this bus.
+
+        Example:
+
+            # wait up to 3 seconds for 10 foo:bar events...
+
+            waiter = bus.waiter(10,'foo:bar')
+
+            # .. fire thread that will cause foo:bar events
+
+            events = waiter.wait(timeout=3)
+
+            if events == None:
+                # handle the timout case...
+
+            for event in events:
+                # parse the events if you need...
+
+        NOTE: use with caution... it's easy to accidentally construct
+              race conditions with this mechanism ;)
+
+        '''
+        return Waiter(self, count, self.loop, *names)
+
+    async def log(self, level, mesg, **info):
```

Comment:
We've never used this in production for anything meaningful. @invisig0th thoughts on gutting it?

Reply:
I'll comment it out.